pkg/rsstorage/servers/s3server/s3_enumeration.go

```diff
@@ -157,6 +158,114 @@ func (a *DefaultAwsOps) BucketObjects(bucket, s3Prefix string, concurrency int,
 	}
 }
 
+func (a *DefaultAwsOps) BucketObjectsMap(bucket, s3Prefix string, concurrency int, recursive bool, reg *regexp.Regexp) (map[string]string, error) {
+	svc := s3.New(a.sess)
+
+	nextMarkerChan := make(chan string, 100)
+	nextMarkerChan <- ""
+	defer close(nextMarkerChan)
+
+	binaryMeta := make(map[string]string)
+	binaryL := sync.Mutex{}
+
+	wg := sync.WaitGroup{}
+	waitCh := make(chan struct{})
+	wg.Add(1)
+
+	var ops uint64
+	var total uint64
+
+	errCh := make(chan error)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// If recursive is not true, include a delimiter so we only list the contents
+	// of the directory indicated by `s3Prefix`. Otherwise, leave the delimiter nil
+	// so we list everything recursively.
+	var delimiter *string
+	if !recursive {
+		delimiter = aws.String("/")
+	}
+
+	go func() {
+		for i := 0; i < concurrency; i++ {
+			go func() {
+				for nextMarker := range nextMarkerChan {
+					wg.Add(1)
+
+					query := &s3.ListObjectsInput{
+						Bucket:    aws.String(bucket),
+						Prefix:    aws.String(s3Prefix),
+						Delimiter: delimiter,
+					}
+
+					if nextMarker != "" {
+						query.SetMarker(nextMarker)
+					}
+
+					resp, err := svc.ListObjectsWithContext(ctx, query)
+					if err != nil {
+						errCh <- fmt.Errorf("something went wrong listing objects: %s", err)
+						return
+					}
+
+					nm := ""
+
+					if resp.NextMarker != nil {
+						nm = *resp.NextMarker
+						nextMarkerChan <- nm
+					}
+
+					// When there are no contents, we need to return
+					// early.
+					if len(resp.Contents) == 0 {
+						wg.Done()
+						// TODO: `nm` may always be blank when there are no
+						// contents, so this conditional may be unnecessary.
+						if nm == "" {
+							wg.Done()
+						}
+						return
+					}
+
+					bm := getObjectsAllMap(resp, s3Prefix, reg)
```

Comment:
This is the only line that differs materially from `BucketObjects`. It calls the new `getObjectsAllMap` method instead of `getObjectsAll`.